### ISSUE TITLE
Add unit tests

### ISSUE THREAD

**Issue by andreeaflorescu** [rust-vmm/vfio-ioctls#39]
This crate does not have any test. We discussed about this at the sync meeting and @jiangliu mentioned that there are some hardware dependencies.

Can we come up with a list of tests that we can implement that could work on the current host running the CI?

Something really basic works on my machine, so I assume it will work in the CI as well:
```rust
#[cfg(test)]
mod test {
    use std::sync::Arc;
    use kvm_bindings::bindings::{kvm_create_device, kvm_device_type_KVM_DEV_TYPE_VFIO};
    use kvm_ioctls::Kvm;
    use crate::{VfioContainer, VfioDevice};

    #[test]
    fn test_create_container() {
        let kvm = Kvm::new().unwrap();
        let vm = kvm.create_vm().unwrap();

        let mut vfio_dev = kvm_create_device {
            type_: kvm_device_type_KVM_DEV_TYPE_VFIO,
            fd: 0,
            flags: 0,
        };

        let device_fd = vm.create_device(&mut vfio_dev).unwrap();
        let vfio_container = VfioContainer::new(Arc::new(device_fd)).unwrap();
    }
}
```

At this point this test is not doing anything useful, but we can add some checks about assumptions that need to hold true. I would be interested in coming up with a list of test that we can write so maybe we can add them in separate issues.

@jiangliu @rbradford @sboeuf what do you think?

**Comment by sboeuf** [comment]
Well this is gonna be very limited, as I don't think you'll be able to write/run many other tests. You're gonna be blocked very quickly by the fact you don't really have a device.
But I still agree it's important to create testing for this crate, so we need a way to have access to a dedicated device that would be bound to VFIO on the machine the unit test is run. We'd need to make sure the unit test run in sequence and not in parallel to avoid issues with the device being opened multiple times (which won't work). We would also have to be okay with the fact that these unit tests require a special environment and they can be run exclusively on the CI machines, as we would know the device is there.
Another solution would be to run these tests inside a virtual machine as we could decide to create a dedicated virtio device to be attached behind a vIOMMU so that it could be bound to VFIO inside the guest. And from there, you would run the unit tests relying on this virtio device that you know about. This would allow for advanced testing as you could check you get the expected regions and info from the device.

**Comment by andreeaflorescu** [comment]
All the tests are already running in the rust-vmm-container. Is there something that we can add to that container to make this possible? What we did for other crates was to hide the tests under a feature when the tests should not be run on any machine (because for example they're using a lot of memory). We can think about a similar approach here. To be able to run the tests sequentially, we need to do some updates in rust-vmm-ci, but that should not be a problem.

**Comment by sboeuf** [comment]
Oh the part about running these tests conditionally is quite simple. We usually create a Rust namespace for each special type of test. And for the sequential part, it's only about running the tests with one thread.
The hard part is about deciding what kind of approach we pick to get access to a PCI device that can be used for these tests. If you can make sure the CI machine can always expose a dedicated PCI device like a NIC for instance, we can simply have a bit of scripting before the CI runs to make sure the device is bound to the vfio-pci driver and afterwards we can have the test running based on this VFIO device.

**Comment by andreeaflorescu** [comment]
We can make some assumptions about the CI host because we're already selecting host based on some properties (i.e. OS = linux/windows, platform = x86/arm). We can introduce a new property for the host to say that it should also have (device=/something/we/want). This will make it such that the tests always run on the instances that have that particular device. For now it would be best to just select some device that is already available on the CI instance, so we don't need to build a new one.

What are some devices that are okay to use in these tests? And how can I make sure they exist on the host?

**Comment by sboeuf** [comment]
Any PCI device would do really. But it needs to sit behind an IOMMU, which might not be the case. You can look at `/sys/kernel/iommu_groups` to check if you have such device. A network card is usually a good fit.

**Comment by jiangliu** [comment]
I'm working on a mock framework for the vfio-ioctls crate, it may help to improve code coverage though it doesn't really talk with hardwares. I have also investigated sample vfio-mdev drivers, but current vfio-ioctls only supports PCI devices:(

**Comment by jiangliu** [comment]
With two-days tweaking, the vfio-ioctls itself reaches 80.5% code coverage now.
https://github.com/rust-vmm/vfio-ioctls/pull/43

**Comment by jiangliu** [comment]
Closed via #44
